From d477007d1e1e466d389307c57b7c82f2d1714d2d Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Tue, 12 Mar 2024 05:12:04 +0200
Subject: [PATCH] [FSW-402] fs1: bootrom_init: clear status registers

Bootrom needs to clear all status registers from
previous boot before starting current boot. This
patch adds function boot_rom_clear_status to clear
all relevant registers.

Change-Id: I259ac85c26c9a600f62c624c4b203095c08b9163
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 7c5bdf7..a95cf2d 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -28,4 +28,23 @@
 		return IO_DIE_SLAVE;
 	else
 		return INVALID_DIE_TYPE;
-}
\ No newline at end of file
+}
+
+void boot_rom_clear_status(void)
+{
+	/* As per Architecture recommendation, clear status reg */
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_STATUS));
+
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_0));
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_1));
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_2));
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_3));
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_4));
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_5));
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_6));
+	regmap_writel(0, IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_7));
+
+	/* Die indication register is on io die slave and compute */
+	if (boot_rom_global.die_type != IO_DIE_MASTER)
+		regmap_writel(0, IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_SCRATCHPAD_2));
+}
diff --git a/common/common.c b/common/common.c
index 75b30b3..c38ebc1 100644
--- a/common/common.c
+++ b/common/common.c
@@ -18,4 +18,6 @@
 	boot_rom_global.die_id = die_id;
 	boot_rom_global.die_type = boot_rom_get_die_type();
 
+	/* clear boot rom Status registers */
+	boot_rom_clear_status();
 }
diff --git a/common/common.h b/common/common.h
new file mode 100644
index 0000000..0dc54ec
--- /dev/null
+++ b/common/common.h
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+#ifndef _COMMON_H_
+#define _COMMON_H_
+
+void boot_rom_clear_status(void);
+
+#endif /* _COMMON_H_ */
